<!doctype html><html lang=en dir=auto><head><meta charset=utf-8><meta http-equiv=x-ua-compatible content="IE=edge"><meta name=viewport content="width=device-width,initial-scale=1,shrink-to-fit=no"><meta name=robots content="index, follow"><title># (Java) 예외처리를 위한 Try-Catch | cjlee38</title><meta name=keywords content="null"><meta name=description content="0. 들어가며 : 예외처리는, 그 이름만 들어도 두려움이 생긴다.
간단하게 생각하자면 if else문의 연장선인 것 같으면서도,
또 뭔가 시스템의 장애를 덕지덕지 막아내는 것 같은 찝찝함도 동시에 든다.
예외처리문을 쓸때면 항상
&ldquo;애초에 코드를 잘 짜면 되지 않나?&rdquo; 라는
바보같은 의문이 남기도 한다.
아무튼, 오늘 다루고자 하는 내용은
&ldquo;예외 처리는 이렇게 해야한다!&rdquo; 하는 거창한 내용은 못되고(쓸 실력도 안되고),
기본적으로 알고 있었던, 그리고 글을 쓰면서 새로 배운,
이러이러한 것들이 존재한다~ 정도의 Introduction 느낌만 가져가려고 한다."><meta name=author content="cjlee38"><link rel=canonical href=http://cjlee38.github.io/post/language/java/2020-10-05-exception_in_java/><link crossorigin=anonymous href=/assets/css/stylesheet.min.48a18943c2fc15c38a372b8dde1f5e5dc0bc64fa6cb90f5a817d2f8c76b7f3ae.css integrity="sha256-SKGJQ8L8FcOKNyuN3h9eXcC8ZPpsuQ9agX0vjHa3864=" rel="preload stylesheet" as=style><script defer crossorigin=anonymous src=/assets/js/highlight.min.4dcb3c4f38462f66c6b6137227726f5543cb934cca9788f041c087e374491df2.js integrity="sha256-Tcs8TzhGL2bGthNyJ3JvVUPLk0zKl4jwQcCH43RJHfI=" onload=hljs.initHighlightingOnLoad()></script>
<link rel=icon href=http://cjlee38.github.io/favicon.ico><link rel=icon type=image/png sizes=16x16 href=http://cjlee38.github.io/favicon-16x16.png><link rel=icon type=image/png sizes=32x32 href=http://cjlee38.github.io/favicon-32x32.png><link rel=apple-touch-icon href=http://cjlee38.github.io/apple-touch-icon.png><link rel=mask-icon href=http://cjlee38.github.io/safari-pinned-tab.svg><meta name=theme-color content="#2e2e33"><meta name=msapplication-TileColor content="#2e2e33"><noscript><style>#theme-toggle,.top-link{display:none}</style></noscript><script type=application/javascript>var doNotTrack=!1;doNotTrack||(function(e,o,i,a,t,n,s){e.GoogleAnalyticsObject=t,e[t]=e[t]||function(){(e[t].q=e[t].q||[]).push(arguments)},e[t].l=1*new Date,n=o.createElement(i),s=o.getElementsByTagName(i)[0],n.async=1,n.src=a,s.parentNode.insertBefore(n,s)}(window,document,"script","https://www.google-analytics.com/analytics.js","ga"),ga("create","UA-174927148-1","auto"),ga("send","pageview"))</script><meta property="og:title" content="# (Java) 예외처리를 위한 Try-Catch"><meta property="og:description" content="0. 들어가며 : 예외처리는, 그 이름만 들어도 두려움이 생긴다.
간단하게 생각하자면 if else문의 연장선인 것 같으면서도,
또 뭔가 시스템의 장애를 덕지덕지 막아내는 것 같은 찝찝함도 동시에 든다.
예외처리문을 쓸때면 항상
&ldquo;애초에 코드를 잘 짜면 되지 않나?&rdquo; 라는
바보같은 의문이 남기도 한다.
아무튼, 오늘 다루고자 하는 내용은
&ldquo;예외 처리는 이렇게 해야한다!&rdquo; 하는 거창한 내용은 못되고(쓸 실력도 안되고),
기본적으로 알고 있었던, 그리고 글을 쓰면서 새로 배운,
이러이러한 것들이 존재한다~ 정도의 Introduction 느낌만 가져가려고 한다."><meta property="og:type" content="article"><meta property="og:url" content="http://cjlee38.github.io/post/language/java/2020-10-05-exception_in_java/"><meta property="article:section" content="post"><meta property="article:published_time" content="2020-10-05T06:42:00+00:00"><meta property="article:modified_time" content="2020-10-05T06:42:00+00:00"><meta property="og:site_name" content="cjlee38"><meta name=twitter:card content="summary"><meta name=twitter:title content="# (Java) 예외처리를 위한 Try-Catch"><meta name=twitter:description content="0. 들어가며 : 예외처리는, 그 이름만 들어도 두려움이 생긴다.
간단하게 생각하자면 if else문의 연장선인 것 같으면서도,
또 뭔가 시스템의 장애를 덕지덕지 막아내는 것 같은 찝찝함도 동시에 든다.
예외처리문을 쓸때면 항상
&ldquo;애초에 코드를 잘 짜면 되지 않나?&rdquo; 라는
바보같은 의문이 남기도 한다.
아무튼, 오늘 다루고자 하는 내용은
&ldquo;예외 처리는 이렇게 해야한다!&rdquo; 하는 거창한 내용은 못되고(쓸 실력도 안되고),
기본적으로 알고 있었던, 그리고 글을 쓰면서 새로 배운,
이러이러한 것들이 존재한다~ 정도의 Introduction 느낌만 가져가려고 한다."><script type=application/ld+json>{"@context":"https://schema.org","@type":"BreadcrumbList","itemListElement":[{"@type":"ListItem","position":1,"name":"Posts","item":"http://cjlee38.github.io/post/"},{"@type":"ListItem","position":2,"name":"# (Java) 예외처리를 위한 Try-Catch","item":"http://cjlee38.github.io/post/language/java/2020-10-05-exception_in_java/"}]}</script><script type=application/ld+json>{"@context":"https://schema.org","@type":"BlogPosting","headline":"# (Java) 예외처리를 위한 Try-Catch","name":"# (Java) 예외처리를 위한 Try-Catch","description":"0. 들어가며 : 예외처리는, 그 이름만 들어도 두려움이 생긴다.\n간단하게 생각하자면 if else문의 연장선인 것 같으면서도,\n또 뭔가 시스템의 장애를 덕지덕지 막아내는 것 같은 찝찝함도 동시에 든다.\n예외처리문을 쓸때면 항상\n\u0026ldquo;애초에 코드를 잘 짜면 되지 않나?\u0026rdquo; 라는\n바보같은 의문이 남기도 한다.\n아무튼, 오늘 다루고자 하는 내용은\n\u0026ldquo;예외 처리는 이렇게 해야한다!\u0026rdquo; 하는 거창한 내용은 못되고(쓸 실력도 안되고),\n기본적으로 알고 있었던, 그리고 글을 쓰면서 새로 배운,\n이러이러한 것들이 존재한다~ 정도의 Introduction 느낌만 가져가려고 한다.","keywords":["null"],"articleBody":"0. 들어가며 : 예외처리는, 그 이름만 들어도 두려움이 생긴다.\n간단하게 생각하자면 if else문의 연장선인 것 같으면서도,\n또 뭔가 시스템의 장애를 덕지덕지 막아내는 것 같은 찝찝함도 동시에 든다.\n예외처리문을 쓸때면 항상\n“애초에 코드를 잘 짜면 되지 않나?” 라는\n바보같은 의문이 남기도 한다.\n아무튼, 오늘 다루고자 하는 내용은\n“예외 처리는 이렇게 해야한다!” 하는 거창한 내용은 못되고(쓸 실력도 안되고),\n기본적으로 알고 있었던, 그리고 글을 쓰면서 새로 배운,\n이러이러한 것들이 존재한다~ 정도의 Introduction 느낌만 가져가려고 한다.\n1. 예외란 무엇인가? : Error와 Exception은 다르다.\n- Error : 시스템 레벨에서 발생하는 심각한 수준의 오류. 프로그램 내에서 처리할 수 없음.\ne.g. OutOfMemoryError, StackOverflowError\n- Exception : 개발자의 설계 미숙 등으로 인해, 로직 상의 문제. 미리 예측하고, 처리할 수 있다.\ne.g. NullPointerException, IndexOutOfBoundException\n1. try - catch - finally : 아마 try-catch 문을 배웠다면,\n가장 먼저 배우는 기본적인 형태는 다음과 같을 것이다.\npublic class Main {   public static void main(String[] args) {  myException();  }   public static void myException() {  try {  System.out.println(\"try here\"); // throw new IllegalStateException(\"exception here\");  } catch(Exception e) {  System.out.println(\"catch here\");  System.out.println(\"e : \" + e.getMessage());  } finally {  System.out.println(\"finally here\");  }  }  } 중간에 주석처리된 throw new … 는 예외를 발생시키는 코드다.\n저 키워드에 대해서는 후술하겠다.\n어쨌든, myException() 이라는 함수가 실행되면,\ntry 구문에서 try here를 출력하고,\n에러가 발생한다면(주석이 해제된다면) catch here 와 e : exception here 를,\n그리고 마지막으로 finally의 finally here를 출력하고 함수가 종료된다.\n만약 이러한 exception에 대한 처리를 해주지 않았더라면,\nNullPointerException, ArrayIndexOutOfBoundsException 등이 발생했을 때\n프로그램이 개복치처럼 죽어버렸을 것이다.\n  Note. 적절한 예외처리에 대한 내용은 잘 모르기도 하고, 이 포스팅의 취지에 맞지 않아\n최대한 빼려고 했는데, 그래도 이 부분은 중요한 것 같아서 꼭 넣어야 할 것 같다.\n무엇이냐 하면, 모든 예외를 Handling 해버리는, 상위 클래스인 Exception으로 예외처리하지 않는 것을 권장한다.\n다시 말해, 특정 예외만을 받아내는 Exception으로 처리해야 한다.\n  또한, finally가 **언제까지 살아있느냐?**를 주의깊게 봐야한다.\n다음의 코드를 보자.\npublic class Main {   public static void main(String[] args) {  myException();  }   public static int myException() {  try {  System.out.println(\"try here\");  throw new IllegalStateException(\"exception here\");  } catch(Exception e) {  System.out.println(\"catch here\");  System.out.println(\"e : \" + e.getMessage());  return -1;  } finally {  System.out.println(\"finally here\");  }  }  } 위에서 작성한 코드와의 차이는 딱 두 가지다.\n try문에서, throw keyword의 주석을 해제했다. myException() 함수의 return을 int로 바꾼 뒤, -1 (임의의 값) 을 return하도록 해서, 이를 출력하도록 했다.  try-catch가 익숙치 않은 사람이라면, finally가 있다 하더라도 return이 있기 때문에\nfinally here는 출력되지 않을 것으로 생각할 수 있지만,\n실제로는 finally here까지 출력이 되고, return 이 이루어진다.\n따라서, 실제 출력은 다음과 같이 이루어진다.\ntry here catch here e : exception here finally here -1 함수를 바로 종료시켜버리는 막강한 return조차 기다리게 만드는 finally라니, 도대체 어떻게 가능한 것일까?\n이를 알아보기 위해, 위 코드를 컴파일 시킨뒤, IDE를 이용하여 다시 디컴파일 시켜보자.\n// // Source code recreated from a .class file by IntelliJ IDEA // (powered by FernFlower decompiler) // public class Main {  public Main() {  }   public static void main(String[] var0) {  int var1 = myException();  System.out.println(var1);  }   public static int myException() {  byte var1;  try {  System.out.println(\"try here\");  throw new IllegalStateException(\"exception here\");  } catch (Exception var5) {  System.out.println(\"catch here\");  System.out.println(\"e : \" + var5.getMessage());  var1 = -1;  } finally {  System.out.println(\"finally here\");  }   return var1;  } } 디컴파일 시킨 코드를 보니, 우리가 작성했던 return 문이 catch 안에서 실행되는 것이 아니고,\nvar1 이라는 byte 변수에 담아 놓은 뒤에, finally 이후에 return 되는 것을 볼 수 있다.\n즉, 우리가 catch안에 return을 작성하더라도, 컴파일러에 의해 뒤로 미뤄진다.\n2. Exception class : 예외 클래스는 다음과 같은 구조를 가진다.\n(인터페이스의 네이밍을 가진) Throwable이라는 클래스 하위에,\n아까 보았던 Exception과 Error가 구분되고,\n그 하위에 Exception을 직접 상속하는 클래스들,\n그리고 Exception을 상속하는 RunTimeException을 상속하는 클래스들로 이루어진다.\n이 때, Checked와 Unchecked는 다음과 같은 차이가 있다.\n- Checked Exception : 컴파일시에 “에러를 처리하는지\"의 여부를 확인한다.\n예외 발생 시 transaction을 roll-back 하지 않는다.\n- Unchecked Exception : 예외가 런타임시에 발생할수도, 발생하지 않을수도 있으므로,\n컴파일할 때에는 에러 처리 여부를 확인하지 않는다.\n예외 발생 시 transaction을 roll-back 한다.\n3. throw와 throws : 이 두가지 키워드가 수행하는 역할은 다르지만, 비슷하게 생겨서 묶었다.\n- throw : 위에서 적어놓았던 throw new IllegalStateException(); 은,\nIllegalStateException 이라는 예외를 “발생\"시킨다.\n즉, throw는 예외를 발생시키는 역할을 한다.\n- throws : 반면, throws는 예외를 던진다. 무슨말인가 하면, 예외가 발생했을 때 이를 메소드 자기 자신이 처리하지 않고,\n자기 자신을 호출한 메소드에게 “이거 처리해주세요” 하고 던진다.\n계속해서 stack을 타고 가다보면, 최종적으로는 main 메소드를 실행한 JVM 에게 던진다.\n다음의 코드의 흐름을 잘 살펴보면, 위 두 키워드의 차이에 대해서 알 수 있으리라 생각한다.\npublic class Main {  public static void main(String[] args) {  try {  myException();  } catch(IllegalStateException e) {  System.out.println(e.getMessage());  }   }   public static int myException() throws IllegalStateException {  throw new IllegalStateException(\"exception\");  } } 4. try-with-resources : 버퍼 입출력을 지원하는 클래스인 BufferedWriter는 Checked Exception중 하나이다.\nPath file = Paths.get(\" ... \"); BufferedWriter writer = null;  writer = Files.newBufferedWriter(file); // IOException writer.write(\" ... \"); // IOException writer.close(); // IOException 위 세 곳에서 IOException이 발생하므로,\nIDE에서 Unhandled Exception이 있다면서 빨간 줄을 쭉 그어준다.\n그러나, close()의 경우 반드시 실행되어야 하기 때문에,\ntry가 아닌 finally block에 들어가야 하는데, 이 또한 IOException이 발생할 수 있다.\n따라서, finally에 try-catch를 한번 더 써야하는데, 이는 꽤 귀찮은 일이다.\n이를 위해, try-with-resources 라는 것이 생겼다.\n형태는 다음과 같다.\ntry(BufferedWriter writer = Files.newBufferedWriter(file)) {  writer.write(\" ... \"); } catach(IOException e) {  // do something } 마치, 파이썬에서 Context Manager라고 불리는 with 과 비슷하다.\n파이썬의 Context Manager는 close() 함수를 호출하지 않아도,\nwith 구문이 끝나면 알아서 호출해준다.\nwith open(\"foo.txt\", \"r\") as f:  file = f.read() 아무튼, 이러한 try-with-resources 의 resources가 되기 위해서는\njava.lang.AutoCloseable 인터페이스를 구현해야 한다.\n","wordCount":"885","inLanguage":"en","datePublished":"2020-10-05T06:42:00Z","dateModified":"2020-10-05T06:42:00Z","author":{"@type":"Person","name":"cjlee38"},"mainEntityOfPage":{"@type":"WebPage","@id":"http://cjlee38.github.io/post/language/java/2020-10-05-exception_in_java/"},"publisher":{"@type":"Organization","name":"cjlee38","logo":{"@type":"ImageObject","url":"http://cjlee38.github.io/favicon.ico"}}}</script></head><body class=dark id=top><script>localStorage.getItem("pref-theme")==="light"&&document.body.classList.remove("dark")</script><header class=header><nav class=nav><div class=logo><a href=http://cjlee38.github.io/ accesskey=h title="Home (Alt + H)"><img src=http://cjlee38.github.io/assets/icon.jpeg alt=logo aria-label=logo height=35>Home</a>
<span class=logo-switches><button id=theme-toggle accesskey=t title="(Alt + T)"><svg id="moon" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentcolor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round"><path d="M21 12.79A9 9 0 1111.21 3 7 7 0 0021 12.79z"/></svg><svg id="sun" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentcolor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round"><circle cx="12" cy="12" r="5"/><line x1="12" y1="1" x2="12" y2="3"/><line x1="12" y1="21" x2="12" y2="23"/><line x1="4.22" y1="4.22" x2="5.64" y2="5.64"/><line x1="18.36" y1="18.36" x2="19.78" y2="19.78"/><line x1="1" y1="12" x2="3" y2="12"/><line x1="21" y1="12" x2="23" y2="12"/><line x1="4.22" y1="19.78" x2="5.64" y2="18.36"/><line x1="18.36" y1="5.64" x2="19.78" y2="4.22"/></svg></button></span></div><ul id=menu><li><a href=http://cjlee38.github.io/categories/ title=categories><span>categories</span></a></li><li><a href=http://cjlee38.github.io/tags/ title=tags><span>tags</span></a></li></ul></nav></header><main class=main><article class=post-single><header class=post-header><div class=breadcrumbs><a href=http://cjlee38.github.io/>Home</a>&nbsp;»&nbsp;<a href=http://cjlee38.github.io/post/>Posts</a></div><h1 class=post-title># (Java) 예외처리를 위한 Try-Catch</h1><div class=post-meta><span title="2020-10-05 06:42:00 +0000 UTC">October 5, 2020</span>&nbsp;·&nbsp;5 min&nbsp;·&nbsp;cjlee38&nbsp;|&nbsp;<a href=https://github.com/%3cpath_to_repo%3e/content/post/Language/Java/2020-10-05-exception_in_java.md rel="noopener noreferrer" target=_blank>Suggest Changes</a></div></header><div class=toc><details open><summary accesskey=c title="(Alt + C)"><span class=details>Table of Contents</span></summary><div class=inner><ul><li><a href=#0-%eb%93%a4%ec%96%b4%ea%b0%80%eb%a9%b0 aria-label="0. 들어가며">0. 들어가며</a></li><li><a href=#1-%ec%98%88%ec%99%b8%eb%9e%80-%eb%ac%b4%ec%97%87%ec%9d%b8%ea%b0%80 aria-label="1. 예외란 무엇인가?">1. 예외란 무엇인가?</a><ul><ul><li><a href=#--error aria-label="- Error">- Error</a></li><li><a href=#--exception aria-label="- Exception">- Exception</a></li></ul></ul></li><li><a href=#1-try---catch---finally aria-label="1. try - catch - finally">1. try - catch - finally</a></li><li><a href=#2-exception-class aria-label="2. Exception class">2. Exception class</a><ul><ul><li><a href=#--checked-exception aria-label="- Checked Exception">- Checked Exception</a></li><li><a href=#--unchecked-exception aria-label="- Unchecked Exception">- Unchecked Exception</a></li></ul></ul></li><li><a href=#3-throw%ec%99%80-throws aria-label="3. throw와 throws">3. throw와 throws</a><ul><ul><li><a href=#--throw aria-label="- throw">- throw</a></li><li><a href=#--throws aria-label="- throws">- throws</a></li></ul></ul></li><li><a href=#4-try-with-resources aria-label="4. try-with-resources">4. try-with-resources</a></li></ul></div></details></div><div class=post-content><h1 id=0-들어가며>0. 들어가며<a hidden class=anchor aria-hidden=true href=#0-들어가며>#</a></h1><p>: 예외처리는, 그 이름만 들어도 두려움이 생긴다.<br>간단하게 생각하자면 if else문의 연장선인 것 같으면서도,<br>또 뭔가 시스템의 장애를 덕지덕지 막아내는 것 같은 찝찝함도 동시에 든다.</p><p>예외처리문을 쓸때면 항상<br>&ldquo;애초에 코드를 잘 짜면 되지 않나?&rdquo; 라는<br>바보같은 의문이 남기도 한다.</p><p>아무튼, 오늘 다루고자 하는 내용은<br>&ldquo;예외 처리는 이렇게 해야한다!&rdquo; 하는 거창한 내용은 못되고(쓸 실력도 안되고),<br>기본적으로 알고 있었던, 그리고 글을 쓰면서 새로 배운,<br>이러이러한 것들이 존재한다~ 정도의 Introduction 느낌만 가져가려고 한다.</p><h1 id=1-예외란-무엇인가>1. 예외란 무엇인가?<a hidden class=anchor aria-hidden=true href=#1-예외란-무엇인가>#</a></h1><p>: Error와 Exception은 다르다.</p><h3 id=--error>- Error<a hidden class=anchor aria-hidden=true href=#--error>#</a></h3><p>: 시스템 레벨에서 발생하는 <strong>심각한</strong> 수준의 오류. 프로그램 내에서 처리할 수 없음.<br>e.g. OutOfMemoryError, StackOverflowError</p><h3 id=--exception>- Exception<a hidden class=anchor aria-hidden=true href=#--exception>#</a></h3><p>: 개발자의 설계 미숙 등으로 인해, 로직 상의 문제. 미리 예측하고, 처리할 수 있다.<br>e.g. NullPointerException, IndexOutOfBoundException</p><h1 id=1-try---catch---finally>1. try - catch - finally<a hidden class=anchor aria-hidden=true href=#1-try---catch---finally>#</a></h1><p>: 아마 try-catch 문을 배웠다면,<br>가장 먼저 배우는 기본적인 형태는 다음과 같을 것이다.</p><div class=highlight><pre tabindex=0 style=color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-java data-lang=java><span style=display:flex><span><span style=color:#66d9ef>public</span> <span style=color:#66d9ef>class</span> <span style=color:#a6e22e>Main</span> <span style=color:#f92672>{</span>
</span></span><span style=display:flex><span>
</span></span><span style=display:flex><span>    <span style=color:#66d9ef>public</span> <span style=color:#66d9ef>static</span> <span style=color:#66d9ef>void</span> <span style=color:#a6e22e>main</span><span style=color:#f92672>(</span>String<span style=color:#f92672>[]</span> args<span style=color:#f92672>)</span> <span style=color:#f92672>{</span>
</span></span><span style=display:flex><span>        myException<span style=color:#f92672>();</span>
</span></span><span style=display:flex><span>    <span style=color:#f92672>}</span>
</span></span><span style=display:flex><span>
</span></span><span style=display:flex><span>    <span style=color:#66d9ef>public</span> <span style=color:#66d9ef>static</span> <span style=color:#66d9ef>void</span> <span style=color:#a6e22e>myException</span><span style=color:#f92672>()</span> <span style=color:#f92672>{</span>
</span></span><span style=display:flex><span>        <span style=color:#66d9ef>try</span> <span style=color:#f92672>{</span>
</span></span><span style=display:flex><span>            System<span style=color:#f92672>.</span><span style=color:#a6e22e>out</span><span style=color:#f92672>.</span><span style=color:#a6e22e>println</span><span style=color:#f92672>(</span><span style=color:#e6db74>&#34;try here&#34;</span><span style=color:#f92672>);</span>
</span></span><span style=display:flex><span><span style=color:#75715e>//            throw new IllegalStateException(&#34;exception here&#34;);
</span></span></span><span style=display:flex><span><span style=color:#75715e></span>        <span style=color:#f92672>}</span> <span style=color:#66d9ef>catch</span><span style=color:#f92672>(</span>Exception e<span style=color:#f92672>)</span> <span style=color:#f92672>{</span>
</span></span><span style=display:flex><span>            System<span style=color:#f92672>.</span><span style=color:#a6e22e>out</span><span style=color:#f92672>.</span><span style=color:#a6e22e>println</span><span style=color:#f92672>(</span><span style=color:#e6db74>&#34;catch here&#34;</span><span style=color:#f92672>);</span>
</span></span><span style=display:flex><span>            System<span style=color:#f92672>.</span><span style=color:#a6e22e>out</span><span style=color:#f92672>.</span><span style=color:#a6e22e>println</span><span style=color:#f92672>(</span><span style=color:#e6db74>&#34;e : &#34;</span> <span style=color:#f92672>+</span> e<span style=color:#f92672>.</span><span style=color:#a6e22e>getMessage</span><span style=color:#f92672>());</span>
</span></span><span style=display:flex><span>        <span style=color:#f92672>}</span> <span style=color:#66d9ef>finally</span> <span style=color:#f92672>{</span>
</span></span><span style=display:flex><span>            System<span style=color:#f92672>.</span><span style=color:#a6e22e>out</span><span style=color:#f92672>.</span><span style=color:#a6e22e>println</span><span style=color:#f92672>(</span><span style=color:#e6db74>&#34;finally here&#34;</span><span style=color:#f92672>);</span>
</span></span><span style=display:flex><span>        <span style=color:#f92672>}</span>
</span></span><span style=display:flex><span>    <span style=color:#f92672>}</span>
</span></span><span style=display:flex><span>
</span></span><span style=display:flex><span><span style=color:#f92672>}</span>
</span></span></code></pre></div><p>중간에 주석처리된 throw new &mldr; 는 예외를 발생시키는 코드다.<br>저 키워드에 대해서는 후술하겠다.</p><p>어쨌든, myException() 이라는 함수가 실행되면,<br>try 구문에서 <code>try here</code>를 출력하고,<br>에러가 발생한다면(<strong>주석이 해제된다면</strong>) <code>catch here</code> 와 <code>e : exception here</code> 를,<br>그리고 마지막으로 finally의 <code>finally here</code>를 출력하고 함수가 종료된다.</p><p>만약 이러한 exception에 대한 처리를 해주지 않았더라면,<br>NullPointerException, ArrayIndexOutOfBoundsException 등이 발생했을 때<br>프로그램이 개복치처럼 죽어버렸을 것이다.</p><hr><blockquote><p>Note. 적절한 예외처리에 대한 내용은 잘 모르기도 하고, 이 포스팅의 취지에 맞지 않아<br>최대한 빼려고 했는데, 그래도 이 부분은 중요한 것 같아서 꼭 넣어야 할 것 같다.<br>무엇이냐 하면, <strong>모든 예외를 Handling 해버리는, 상위 클래스인 Exception으로 예외처리하지 않는 것을 권장한다.</strong><br>다시 말해, <strong>특정 예외만을 받아내는 Exception으로 처리해야 한다.</strong></p></blockquote><hr><p>또한, finally가 **언제까지 살아있느냐?**를 주의깊게 봐야한다.<br>다음의 코드를 보자.</p><div class=highlight><pre tabindex=0 style=color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-java data-lang=java><span style=display:flex><span><span style=color:#66d9ef>public</span> <span style=color:#66d9ef>class</span> <span style=color:#a6e22e>Main</span> <span style=color:#f92672>{</span>
</span></span><span style=display:flex><span>
</span></span><span style=display:flex><span>    <span style=color:#66d9ef>public</span> <span style=color:#66d9ef>static</span> <span style=color:#66d9ef>void</span> <span style=color:#a6e22e>main</span><span style=color:#f92672>(</span>String<span style=color:#f92672>[]</span> args<span style=color:#f92672>)</span> <span style=color:#f92672>{</span>
</span></span><span style=display:flex><span>        myException<span style=color:#f92672>();</span>
</span></span><span style=display:flex><span>    <span style=color:#f92672>}</span>
</span></span><span style=display:flex><span>
</span></span><span style=display:flex><span>    <span style=color:#66d9ef>public</span> <span style=color:#66d9ef>static</span> <span style=color:#66d9ef>int</span> <span style=color:#a6e22e>myException</span><span style=color:#f92672>()</span> <span style=color:#f92672>{</span>
</span></span><span style=display:flex><span>        <span style=color:#66d9ef>try</span> <span style=color:#f92672>{</span>
</span></span><span style=display:flex><span>            System<span style=color:#f92672>.</span><span style=color:#a6e22e>out</span><span style=color:#f92672>.</span><span style=color:#a6e22e>println</span><span style=color:#f92672>(</span><span style=color:#e6db74>&#34;try here&#34;</span><span style=color:#f92672>);</span>
</span></span><span style=display:flex><span>            <span style=color:#66d9ef>throw</span> <span style=color:#66d9ef>new</span> IllegalStateException<span style=color:#f92672>(</span><span style=color:#e6db74>&#34;exception here&#34;</span><span style=color:#f92672>);</span>
</span></span><span style=display:flex><span>        <span style=color:#f92672>}</span> <span style=color:#66d9ef>catch</span><span style=color:#f92672>(</span>Exception e<span style=color:#f92672>)</span> <span style=color:#f92672>{</span>
</span></span><span style=display:flex><span>            System<span style=color:#f92672>.</span><span style=color:#a6e22e>out</span><span style=color:#f92672>.</span><span style=color:#a6e22e>println</span><span style=color:#f92672>(</span><span style=color:#e6db74>&#34;catch here&#34;</span><span style=color:#f92672>);</span>
</span></span><span style=display:flex><span>            System<span style=color:#f92672>.</span><span style=color:#a6e22e>out</span><span style=color:#f92672>.</span><span style=color:#a6e22e>println</span><span style=color:#f92672>(</span><span style=color:#e6db74>&#34;e : &#34;</span> <span style=color:#f92672>+</span> e<span style=color:#f92672>.</span><span style=color:#a6e22e>getMessage</span><span style=color:#f92672>());</span>
</span></span><span style=display:flex><span>            <span style=color:#66d9ef>return</span> <span style=color:#f92672>-</span>1<span style=color:#f92672>;</span>
</span></span><span style=display:flex><span>        <span style=color:#f92672>}</span> <span style=color:#66d9ef>finally</span> <span style=color:#f92672>{</span>
</span></span><span style=display:flex><span>            System<span style=color:#f92672>.</span><span style=color:#a6e22e>out</span><span style=color:#f92672>.</span><span style=color:#a6e22e>println</span><span style=color:#f92672>(</span><span style=color:#e6db74>&#34;finally here&#34;</span><span style=color:#f92672>);</span>
</span></span><span style=display:flex><span>        <span style=color:#f92672>}</span>
</span></span><span style=display:flex><span>    <span style=color:#f92672>}</span>
</span></span><span style=display:flex><span>
</span></span><span style=display:flex><span><span style=color:#f92672>}</span>
</span></span></code></pre></div><p>위에서 작성한 코드와의 차이는 딱 두 가지다.</p><ol><li>try문에서, throw keyword의 주석을 해제했다.</li><li>myException() 함수의 return을 int로 바꾼 뒤, -1 <em>(임의의 값)</em> 을 return하도록 해서, 이를 출력하도록 했다.</li></ol><p>try-catch가 익숙치 않은 사람이라면, finally가 있다 하더라도 return이 있기 때문에<br><code>finally here</code>는 출력되지 않을 것으로 생각할 수 있지만,<br>실제로는 <code>finally here</code>까지 출력이 되고, return 이 이루어진다.</p><p>따라서, 실제 출력은 다음과 같이 이루어진다.</p><pre tabindex=0><code>try here
catch here
e : exception here
finally here
-1
</code></pre><p>함수를 바로 종료시켜버리는 막강한 return조차 기다리게 만드는 finally라니, 도대체 어떻게 가능한 것일까?<br>이를 알아보기 위해, 위 코드를 컴파일 시킨뒤, IDE를 이용하여 다시 디컴파일 시켜보자.</p><div class=highlight><pre tabindex=0 style=color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-java data-lang=java><span style=display:flex><span><span style=color:#75715e>//
</span></span></span><span style=display:flex><span><span style=color:#75715e>// Source code recreated from a .class file by IntelliJ IDEA
</span></span></span><span style=display:flex><span><span style=color:#75715e>// (powered by FernFlower decompiler)
</span></span></span><span style=display:flex><span><span style=color:#75715e>//
</span></span></span><span style=display:flex><span><span style=color:#75715e></span><span style=color:#66d9ef>public</span> <span style=color:#66d9ef>class</span> <span style=color:#a6e22e>Main</span> <span style=color:#f92672>{</span>
</span></span><span style=display:flex><span>    <span style=color:#66d9ef>public</span> <span style=color:#a6e22e>Main</span><span style=color:#f92672>()</span> <span style=color:#f92672>{</span>
</span></span><span style=display:flex><span>    <span style=color:#f92672>}</span>
</span></span><span style=display:flex><span>
</span></span><span style=display:flex><span>    <span style=color:#66d9ef>public</span> <span style=color:#66d9ef>static</span> <span style=color:#66d9ef>void</span> <span style=color:#a6e22e>main</span><span style=color:#f92672>(</span>String<span style=color:#f92672>[]</span> var0<span style=color:#f92672>)</span> <span style=color:#f92672>{</span>
</span></span><span style=display:flex><span>        <span style=color:#66d9ef>int</span> var1 <span style=color:#f92672>=</span> myException<span style=color:#f92672>();</span>
</span></span><span style=display:flex><span>        System<span style=color:#f92672>.</span><span style=color:#a6e22e>out</span><span style=color:#f92672>.</span><span style=color:#a6e22e>println</span><span style=color:#f92672>(</span>var1<span style=color:#f92672>);</span>
</span></span><span style=display:flex><span>    <span style=color:#f92672>}</span>
</span></span><span style=display:flex><span>
</span></span><span style=display:flex><span>    <span style=color:#66d9ef>public</span> <span style=color:#66d9ef>static</span> <span style=color:#66d9ef>int</span> <span style=color:#a6e22e>myException</span><span style=color:#f92672>()</span> <span style=color:#f92672>{</span>
</span></span><span style=display:flex><span>        <span style=color:#66d9ef>byte</span> var1<span style=color:#f92672>;</span>
</span></span><span style=display:flex><span>        <span style=color:#66d9ef>try</span> <span style=color:#f92672>{</span>
</span></span><span style=display:flex><span>            System<span style=color:#f92672>.</span><span style=color:#a6e22e>out</span><span style=color:#f92672>.</span><span style=color:#a6e22e>println</span><span style=color:#f92672>(</span><span style=color:#e6db74>&#34;try here&#34;</span><span style=color:#f92672>);</span>
</span></span><span style=display:flex><span>            <span style=color:#66d9ef>throw</span> <span style=color:#66d9ef>new</span> IllegalStateException<span style=color:#f92672>(</span><span style=color:#e6db74>&#34;exception here&#34;</span><span style=color:#f92672>);</span>
</span></span><span style=display:flex><span>        <span style=color:#f92672>}</span> <span style=color:#66d9ef>catch</span> <span style=color:#f92672>(</span>Exception var5<span style=color:#f92672>)</span> <span style=color:#f92672>{</span>
</span></span><span style=display:flex><span>            System<span style=color:#f92672>.</span><span style=color:#a6e22e>out</span><span style=color:#f92672>.</span><span style=color:#a6e22e>println</span><span style=color:#f92672>(</span><span style=color:#e6db74>&#34;catch here&#34;</span><span style=color:#f92672>);</span>
</span></span><span style=display:flex><span>            System<span style=color:#f92672>.</span><span style=color:#a6e22e>out</span><span style=color:#f92672>.</span><span style=color:#a6e22e>println</span><span style=color:#f92672>(</span><span style=color:#e6db74>&#34;e : &#34;</span> <span style=color:#f92672>+</span> var5<span style=color:#f92672>.</span><span style=color:#a6e22e>getMessage</span><span style=color:#f92672>());</span>
</span></span><span style=display:flex><span>            var1 <span style=color:#f92672>=</span> <span style=color:#f92672>-</span>1<span style=color:#f92672>;</span>
</span></span><span style=display:flex><span>        <span style=color:#f92672>}</span> <span style=color:#66d9ef>finally</span> <span style=color:#f92672>{</span>
</span></span><span style=display:flex><span>            System<span style=color:#f92672>.</span><span style=color:#a6e22e>out</span><span style=color:#f92672>.</span><span style=color:#a6e22e>println</span><span style=color:#f92672>(</span><span style=color:#e6db74>&#34;finally here&#34;</span><span style=color:#f92672>);</span>
</span></span><span style=display:flex><span>        <span style=color:#f92672>}</span>
</span></span><span style=display:flex><span>
</span></span><span style=display:flex><span>        <span style=color:#66d9ef>return</span> var1<span style=color:#f92672>;</span>
</span></span><span style=display:flex><span>    <span style=color:#f92672>}</span>
</span></span><span style=display:flex><span><span style=color:#f92672>}</span>
</span></span></code></pre></div><p>디컴파일 시킨 코드를 보니, 우리가 작성했던 return 문이 catch 안에서 실행되는 것이 아니고,<br>var1 이라는 byte 변수에 담아 놓은 뒤에, finally 이후에 return 되는 것을 볼 수 있다.</p><p>즉, 우리가 catch안에 return을 작성하더라도, <strong>컴파일러에 의해 뒤로 미뤄진다.</strong></p><h1 id=2-exception-class>2. Exception class<a hidden class=anchor aria-hidden=true href=#2-exception-class>#</a></h1><p>: 예외 클래스는 다음과 같은 구조를 가진다.</p><p><img loading=lazy src=/assets/images/2020-10-20-19-38-58_2020-10-05-exception_in_java.md.png alt></p><p>(인터페이스의 네이밍을 가진) Throwable이라는 클래스 하위에,<br>아까 보았던 Exception과 Error가 구분되고,<br>그 하위에 Exception을 직접 상속하는 클래스들,<br>그리고 Exception을 상속하는 RunTimeException을 상속하는 클래스들로 이루어진다.</p><p>이 때, Checked와 Unchecked는 다음과 같은 차이가 있다.</p><h3 id=--checked-exception>- Checked Exception<a hidden class=anchor aria-hidden=true href=#--checked-exception>#</a></h3><p>: 컴파일시에 &ldquo;에러를 처리하는지"의 여부를 확인한다.<br>예외 발생 시 transaction을 roll-back 하지 않는다.</p><h3 id=--unchecked-exception>- Unchecked Exception<a hidden class=anchor aria-hidden=true href=#--unchecked-exception>#</a></h3><p>: 예외가 런타임시에 발생할수도, 발생하지 않을수도 있으므로,<br>컴파일할 때에는 에러 처리 여부를 확인하지 않는다.<br>예외 발생 시 transaction을 roll-back 한다.</p><h1 id=3-throw와-throws>3. throw와 throws<a hidden class=anchor aria-hidden=true href=#3-throw와-throws>#</a></h1><p>: 이 두가지 키워드가 수행하는 역할은 다르지만, 비슷하게 생겨서 묶었다.</p><h3 id=--throw>- throw<a hidden class=anchor aria-hidden=true href=#--throw>#</a></h3><p>: 위에서 적어놓았던 <code>throw new IllegalStateException();</code> 은,<br><strong>IllegalStateException 이라는 예외를 &ldquo;발생"시킨다.</strong><br>즉, throw는 예외를 발생시키는 역할을 한다.</p><h3 id=--throws>- throws<a hidden class=anchor aria-hidden=true href=#--throws>#</a></h3><p>: 반면, throws는 예외를 <strong>던진다.</strong>
무슨말인가 하면, 예외가 발생했을 때 이를 메소드 자기 자신이 처리하지 않고,<br>자기 자신을 <strong>호출한</strong> 메소드에게 &ldquo;이거 처리해주세요&rdquo; 하고 던진다.</p><p>계속해서 stack을 타고 가다보면, 최종적으로는 main 메소드를 실행한 JVM 에게 던진다.</p><p>다음의 코드의 흐름을 잘 살펴보면, 위 두 키워드의 차이에 대해서 알 수 있으리라 생각한다.</p><div class=highlight><pre tabindex=0 style=color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-java data-lang=java><span style=display:flex><span><span style=color:#66d9ef>public</span> <span style=color:#66d9ef>class</span> <span style=color:#a6e22e>Main</span> <span style=color:#f92672>{</span>
</span></span><span style=display:flex><span>    <span style=color:#66d9ef>public</span> <span style=color:#66d9ef>static</span> <span style=color:#66d9ef>void</span> <span style=color:#a6e22e>main</span><span style=color:#f92672>(</span>String<span style=color:#f92672>[]</span> args<span style=color:#f92672>)</span> <span style=color:#f92672>{</span>
</span></span><span style=display:flex><span>        <span style=color:#66d9ef>try</span> <span style=color:#f92672>{</span>
</span></span><span style=display:flex><span>            myException<span style=color:#f92672>();</span>
</span></span><span style=display:flex><span>        <span style=color:#f92672>}</span> <span style=color:#66d9ef>catch</span><span style=color:#f92672>(</span>IllegalStateException e<span style=color:#f92672>)</span> <span style=color:#f92672>{</span>
</span></span><span style=display:flex><span>            System<span style=color:#f92672>.</span><span style=color:#a6e22e>out</span><span style=color:#f92672>.</span><span style=color:#a6e22e>println</span><span style=color:#f92672>(</span>e<span style=color:#f92672>.</span><span style=color:#a6e22e>getMessage</span><span style=color:#f92672>());</span>
</span></span><span style=display:flex><span>        <span style=color:#f92672>}</span>
</span></span><span style=display:flex><span>
</span></span><span style=display:flex><span>    <span style=color:#f92672>}</span>
</span></span><span style=display:flex><span>
</span></span><span style=display:flex><span>    <span style=color:#66d9ef>public</span> <span style=color:#66d9ef>static</span> <span style=color:#66d9ef>int</span> <span style=color:#a6e22e>myException</span><span style=color:#f92672>()</span> <span style=color:#66d9ef>throws</span> IllegalStateException <span style=color:#f92672>{</span>
</span></span><span style=display:flex><span>        <span style=color:#66d9ef>throw</span> <span style=color:#66d9ef>new</span> IllegalStateException<span style=color:#f92672>(</span><span style=color:#e6db74>&#34;exception&#34;</span><span style=color:#f92672>);</span>
</span></span><span style=display:flex><span>    <span style=color:#f92672>}</span>
</span></span><span style=display:flex><span><span style=color:#f92672>}</span>
</span></span></code></pre></div><h1 id=4-try-with-resources>4. try-with-resources<a hidden class=anchor aria-hidden=true href=#4-try-with-resources>#</a></h1><p>: 버퍼 입출력을 지원하는 클래스인 BufferedWriter는 Checked Exception중 하나이다.</p><div class=highlight><pre tabindex=0 style=color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-java data-lang=java><span style=display:flex><span>Path file <span style=color:#f92672>=</span>  Paths<span style=color:#f92672>.</span><span style=color:#a6e22e>get</span><span style=color:#f92672>(</span><span style=color:#e6db74>&#34; ... &#34;</span><span style=color:#f92672>);</span>
</span></span><span style=display:flex><span>BufferedWriter writer <span style=color:#f92672>=</span> <span style=color:#66d9ef>null</span><span style=color:#f92672>;</span>
</span></span><span style=display:flex><span>
</span></span><span style=display:flex><span>writer <span style=color:#f92672>=</span> Files<span style=color:#f92672>.</span><span style=color:#a6e22e>newBufferedWriter</span><span style=color:#f92672>(</span>file<span style=color:#f92672>);</span> <span style=color:#75715e>// IOException
</span></span></span><span style=display:flex><span><span style=color:#75715e></span>writer<span style=color:#f92672>.</span><span style=color:#a6e22e>write</span><span style=color:#f92672>(</span><span style=color:#e6db74>&#34; ... &#34;</span><span style=color:#f92672>);</span> <span style=color:#75715e>// IOException
</span></span></span><span style=display:flex><span><span style=color:#75715e></span>writer<span style=color:#f92672>.</span><span style=color:#a6e22e>close</span><span style=color:#f92672>();</span> <span style=color:#75715e>// IOException
</span></span></span></code></pre></div><p>위 세 곳에서 IOException이 발생하므로,<br>IDE에서 Unhandled Exception이 있다면서 빨간 줄을 쭉 그어준다.</p><p>그러나, close()의 경우 <strong>반드시 실행되어야 하기 때문에,</strong><br>try가 아닌 finally block에 들어가야 하는데, 이 또한 IOException이 발생할 수 있다.<br>따라서, finally에 try-catch를 한번 더 써야하는데, 이는 꽤 귀찮은 일이다.</p><p>이를 위해, try-with-resources 라는 것이 생겼다.<br>형태는 다음과 같다.</p><div class=highlight><pre tabindex=0 style=color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-java data-lang=java><span style=display:flex><span><span style=color:#66d9ef>try</span><span style=color:#f92672>(</span>BufferedWriter writer <span style=color:#f92672>=</span> Files<span style=color:#f92672>.</span><span style=color:#a6e22e>newBufferedWriter</span><span style=color:#f92672>(</span>file<span style=color:#f92672>))</span> <span style=color:#f92672>{</span>
</span></span><span style=display:flex><span>    writer<span style=color:#f92672>.</span><span style=color:#a6e22e>write</span><span style=color:#f92672>(</span><span style=color:#e6db74>&#34; ... &#34;</span><span style=color:#f92672>);</span>
</span></span><span style=display:flex><span><span style=color:#f92672>}</span> catach<span style=color:#f92672>(</span>IOException e<span style=color:#f92672>)</span> <span style=color:#f92672>{</span>
</span></span><span style=display:flex><span>    <span style=color:#75715e>// do something
</span></span></span><span style=display:flex><span><span style=color:#75715e></span><span style=color:#f92672>}</span>
</span></span></code></pre></div><p>마치, 파이썬에서 Context Manager라고 불리는 with 과 비슷하다.<br>파이썬의 Context Manager는 close() 함수를 호출하지 않아도,<br>with 구문이 끝나면 알아서 호출해준다.</p><div class=highlight><pre tabindex=0 style=color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-python data-lang=python><span style=display:flex><span><span style=color:#66d9ef>with</span> open(<span style=color:#e6db74>&#34;foo.txt&#34;</span>, <span style=color:#e6db74>&#34;r&#34;</span>) <span style=color:#66d9ef>as</span> f:
</span></span><span style=display:flex><span>    file <span style=color:#f92672>=</span> f<span style=color:#f92672>.</span>read()
</span></span></code></pre></div><p>아무튼, 이러한 try-with-resources 의 resources가 되기 위해서는<br><code>java.lang.AutoCloseable</code> 인터페이스를 구현해야 한다.</p></div><footer class=post-footer><ul class=post-tags><li><a href=http://cjlee38.github.io/tags/null/>null</a></li></ul><nav class=paginav><a class=prev href=http://cjlee38.github.io/post/tech/jsp/2020-10-06-jsp_learning_07/><span class=title>« Prev Page</span><br><span># 7. JSP 학습기록 - MVC 패턴과 JSP Model 1, Model 2</span></a>
<a class=next href=http://cjlee38.github.io/post/algorithm/problem_solving/2020-10-04-problem_solving_12/><span class=title>Next Page »</span><br><span># 프로그래머스[Lv.3] 단속카메라 ( Java )</span></a></nav><div class=share-buttons><a target=_blank rel="noopener noreferrer" aria-label="share # (Java) 예외처리를 위한 Try-Catch on twitter" href="https://twitter.com/intent/tweet/?text=%23%20%28Java%29%20%ec%98%88%ec%99%b8%ec%b2%98%eb%a6%ac%eb%a5%bc%20%ec%9c%84%ed%95%9c%20Try-Catch&url=http%3a%2f%2fcjlee38.github.io%2fpost%2flanguage%2fjava%2f2020-10-05-exception_in_java%2f&hashtags=null"><svg viewBox="0 0 512 512"><path d="M449.446.0C483.971.0 512 28.03 512 62.554v386.892C512 483.97 483.97 512 449.446 512H62.554c-34.524.0-62.554-28.03-62.554-62.554V62.554c0-34.524 28.029-62.554 62.554-62.554h386.892zM195.519 424.544c135.939.0 210.268-112.643 210.268-210.268.0-3.218.0-6.437-.153-9.502 14.406-10.421 26.973-23.448 36.935-38.314-13.18 5.824-27.433 9.809-42.452 11.648 15.326-9.196 26.973-23.602 32.49-40.92-14.252 8.429-30.038 14.56-46.896 17.931-13.487-14.406-32.644-23.295-53.946-23.295-40.767.0-73.87 33.104-73.87 73.87.0 5.824.613 11.494 1.992 16.858-61.456-3.065-115.862-32.49-152.337-77.241-6.284 10.881-9.962 23.601-9.962 37.088.0 25.594 13.027 48.276 32.95 61.456-12.107-.307-23.448-3.678-33.41-9.196v.92c0 35.862 25.441 65.594 59.311 72.49-6.13 1.686-12.72 2.606-19.464 2.606-4.751.0-9.348-.46-13.946-1.38 9.349 29.426 36.628 50.728 68.965 51.341-25.287 19.771-57.164 31.571-91.8 31.571-5.977.0-11.801-.306-17.625-1.073 32.337 21.15 71.264 33.41 112.95 33.41z"/></svg></a><a target=_blank rel="noopener noreferrer" aria-label="share # (Java) 예외처리를 위한 Try-Catch on linkedin" href="https://www.linkedin.com/shareArticle?mini=true&url=http%3a%2f%2fcjlee38.github.io%2fpost%2flanguage%2fjava%2f2020-10-05-exception_in_java%2f&title=%23%20%28Java%29%20%ec%98%88%ec%99%b8%ec%b2%98%eb%a6%ac%eb%a5%bc%20%ec%9c%84%ed%95%9c%20Try-Catch&summary=%23%20%28Java%29%20%ec%98%88%ec%99%b8%ec%b2%98%eb%a6%ac%eb%a5%bc%20%ec%9c%84%ed%95%9c%20Try-Catch&source=http%3a%2f%2fcjlee38.github.io%2fpost%2flanguage%2fjava%2f2020-10-05-exception_in_java%2f"><svg viewBox="0 0 512 512"><path d="M449.446.0C483.971.0 512 28.03 512 62.554v386.892C512 483.97 483.97 512 449.446 512H62.554c-34.524.0-62.554-28.03-62.554-62.554V62.554c0-34.524 28.029-62.554 62.554-62.554h386.892zM160.461 423.278V197.561h-75.04v225.717h75.04zm270.539.0V293.839c0-69.333-37.018-101.586-86.381-101.586-39.804.0-57.634 21.891-67.617 37.266v-31.958h-75.021c.995 21.181.0 225.717.0 225.717h75.02V297.222c0-6.748.486-13.492 2.474-18.315 5.414-13.475 17.767-27.434 38.494-27.434 27.135.0 38.007 20.707 38.007 51.037v120.768H431zM123.448 88.722C97.774 88.722 81 105.601 81 127.724c0 21.658 16.264 39.002 41.455 39.002h.484c26.165.0 42.452-17.344 42.452-39.002-.485-22.092-16.241-38.954-41.943-39.002z"/></svg></a><a target=_blank rel="noopener noreferrer" aria-label="share # (Java) 예외처리를 위한 Try-Catch on reddit" href="https://reddit.com/submit?url=http%3a%2f%2fcjlee38.github.io%2fpost%2flanguage%2fjava%2f2020-10-05-exception_in_java%2f&title=%23%20%28Java%29%20%ec%98%88%ec%99%b8%ec%b2%98%eb%a6%ac%eb%a5%bc%20%ec%9c%84%ed%95%9c%20Try-Catch"><svg viewBox="0 0 512 512"><path d="M449.446.0C483.971.0 512 28.03 512 62.554v386.892C512 483.97 483.97 512 449.446 512H62.554c-34.524.0-62.554-28.03-62.554-62.554V62.554c0-34.524 28.029-62.554 62.554-62.554h386.892zM446 265.638c0-22.964-18.616-41.58-41.58-41.58-11.211.0-21.361 4.457-28.841 11.666-28.424-20.508-67.586-33.757-111.204-35.278l18.941-89.121 61.884 13.157c.756 15.734 13.642 28.29 29.56 28.29 16.407.0 29.706-13.299 29.706-29.701.0-16.403-13.299-29.702-29.706-29.702-11.666.0-21.657 6.792-26.515 16.578l-69.105-14.69c-1.922-.418-3.939-.042-5.585 1.036-1.658 1.073-2.811 2.761-3.224 4.686l-21.152 99.438c-44.258 1.228-84.046 14.494-112.837 35.232-7.468-7.164-17.589-11.591-28.757-11.591-22.965.0-41.585 18.616-41.585 41.58.0 16.896 10.095 31.41 24.568 37.918-.639 4.135-.99 8.328-.99 12.576.0 63.977 74.469 115.836 166.33 115.836s166.334-51.859 166.334-115.836c0-4.218-.347-8.387-.977-12.493 14.564-6.47 24.735-21.034 24.735-38.001zM326.526 373.831c-20.27 20.241-59.115 21.816-70.534 21.816-11.428.0-50.277-1.575-70.522-21.82-3.007-3.008-3.007-7.882.0-10.889 3.003-2.999 7.882-3.003 10.885.0 12.777 12.781 40.11 17.317 59.637 17.317 19.522.0 46.86-4.536 59.657-17.321 3.016-2.999 7.886-2.995 10.885.008 3.008 3.011 3.003 7.882-.008 10.889zm-5.23-48.781c-16.373.0-29.701-13.324-29.701-29.698.0-16.381 13.328-29.714 29.701-29.714 16.378.0 29.706 13.333 29.706 29.714.0 16.374-13.328 29.698-29.706 29.698zM160.91 295.348c0-16.381 13.328-29.71 29.714-29.71 16.369.0 29.689 13.329 29.689 29.71.0 16.373-13.32 29.693-29.689 29.693-16.386.0-29.714-13.32-29.714-29.693z"/></svg></a><a target=_blank rel="noopener noreferrer" aria-label="share # (Java) 예외처리를 위한 Try-Catch on facebook" href="https://facebook.com/sharer/sharer.php?u=http%3a%2f%2fcjlee38.github.io%2fpost%2flanguage%2fjava%2f2020-10-05-exception_in_java%2f"><svg viewBox="0 0 512 512"><path d="M449.446.0C483.971.0 512 28.03 512 62.554v386.892C512 483.97 483.97 512 449.446 512H342.978V319.085h66.6l12.672-82.621h-79.272v-53.617c0-22.603 11.073-44.636 46.58-44.636H425.6v-70.34s-32.71-5.582-63.982-5.582c-65.288.0-107.96 39.569-107.96 111.204v62.971h-72.573v82.621h72.573V512h-191.104c-34.524.0-62.554-28.03-62.554-62.554V62.554c0-34.524 28.029-62.554 62.554-62.554h386.892z"/></svg></a><a target=_blank rel="noopener noreferrer" aria-label="share # (Java) 예외처리를 위한 Try-Catch on whatsapp" href="https://api.whatsapp.com/send?text=%23%20%28Java%29%20%ec%98%88%ec%99%b8%ec%b2%98%eb%a6%ac%eb%a5%bc%20%ec%9c%84%ed%95%9c%20Try-Catch%20-%20http%3a%2f%2fcjlee38.github.io%2fpost%2flanguage%2fjava%2f2020-10-05-exception_in_java%2f"><svg viewBox="0 0 512 512"><path d="M449.446.0C483.971.0 512 28.03 512 62.554v386.892C512 483.97 483.97 512 449.446 512H62.554c-34.524.0-62.554-28.03-62.554-62.554V62.554c0-34.524 28.029-62.554 62.554-62.554h386.892zm-58.673 127.703c-33.842-33.881-78.847-52.548-126.798-52.568-98.799.0-179.21 80.405-179.249 179.234-.013 31.593 8.241 62.428 23.927 89.612l-25.429 92.884 95.021-24.925c26.181 14.28 55.659 21.807 85.658 21.816h.074c98.789.0 179.206-80.413 179.247-179.243.018-47.895-18.61-92.93-52.451-126.81zM263.976 403.485h-.06c-26.734-.01-52.954-7.193-75.828-20.767l-5.441-3.229-56.386 14.792 15.05-54.977-3.542-5.637c-14.913-23.72-22.791-51.136-22.779-79.287.033-82.142 66.867-148.971 149.046-148.971 39.793.014 77.199 15.531 105.329 43.692 28.128 28.16 43.609 65.592 43.594 105.4-.034 82.149-66.866 148.983-148.983 148.984zm81.721-111.581c-4.479-2.242-26.499-13.075-30.604-14.571-4.105-1.495-7.091-2.241-10.077 2.241-2.986 4.483-11.569 14.572-14.182 17.562-2.612 2.988-5.225 3.364-9.703 1.12-4.479-2.241-18.91-6.97-36.017-22.23C231.8 264.15 222.81 249.484 220.198 245s-.279-6.908 1.963-9.14c2.016-2.007 4.48-5.232 6.719-7.847 2.24-2.615 2.986-4.484 4.479-7.472 1.493-2.99.747-5.604-.374-7.846-1.119-2.241-10.077-24.288-13.809-33.256-3.635-8.733-7.327-7.55-10.077-7.688-2.609-.13-5.598-.158-8.583-.158-2.986.0-7.839 1.121-11.944 5.604-4.105 4.484-15.675 15.32-15.675 37.364.0 22.046 16.048 43.342 18.287 46.332 2.24 2.99 31.582 48.227 76.511 67.627 10.685 4.615 19.028 7.371 25.533 9.434 10.728 3.41 20.492 2.929 28.209 1.775 8.605-1.285 26.499-10.833 30.231-21.295 3.732-10.464 3.732-19.431 2.612-21.298-1.119-1.869-4.105-2.99-8.583-5.232z"/></svg></a><a target=_blank rel="noopener noreferrer" aria-label="share # (Java) 예외처리를 위한 Try-Catch on telegram" href="https://telegram.me/share/url?text=%23%20%28Java%29%20%ec%98%88%ec%99%b8%ec%b2%98%eb%a6%ac%eb%a5%bc%20%ec%9c%84%ed%95%9c%20Try-Catch&url=http%3a%2f%2fcjlee38.github.io%2fpost%2flanguage%2fjava%2f2020-10-05-exception_in_java%2f"><svg viewBox="2 2 28 28"><path d="M26.49 29.86H5.5a3.37 3.37.0 01-2.47-1 3.35 3.35.0 01-1-2.47V5.48A3.36 3.36.0 013 3 3.37 3.37.0 015.5 2h21A3.38 3.38.0 0129 3a3.36 3.36.0 011 2.46V26.37a3.35 3.35.0 01-1 2.47 3.38 3.38.0 01-2.51 1.02zm-5.38-6.71a.79.79.0 00.85-.66L24.73 9.24a.55.55.0 00-.18-.46.62.62.0 00-.41-.17q-.08.0-16.53 6.11a.59.59.0 00-.41.59.57.57.0 00.43.52l4 1.24 1.61 4.83a.62.62.0 00.63.43.56.56.0 00.4-.17L16.54 20l4.09 3A.9.9.0 0021.11 23.15zM13.8 20.71l-1.21-4q8.72-5.55 8.78-5.55c.15.0.23.0.23.16a.18.18.0 010 .06s-2.51 2.3-7.52 6.8z"/></svg></a></div></footer></article></main><footer class=footer><span>&copy; 2022 <a href=http://cjlee38.github.io/>cjlee38</a></span>
<span>Powered by
<a href=https://gohugo.io/ rel="noopener noreferrer" target=_blank>Hugo</a> &
        <a href=https://git.io/hugopapermod rel=noopener target=_blank>PaperMod</a></span></footer><a href=#top aria-label="go to top" title="Go to Top (Alt + G)" class=top-link id=top-link accesskey=g><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 12 6" fill="currentcolor"><path d="M12 6H0l6-6z"/></svg></a><script>let menu=document.getElementById("menu");menu&&(menu.scrollLeft=localStorage.getItem("menu-scroll-position"),menu.onscroll=function(){localStorage.setItem("menu-scroll-position",menu.scrollLeft)}),document.querySelectorAll('a[href^="#"]').forEach(e=>{e.addEventListener("click",function(t){t.preventDefault();var e=this.getAttribute("href").substr(1);window.matchMedia("(prefers-reduced-motion: reduce)").matches?document.querySelector(`[id='${decodeURIComponent(e)}']`).scrollIntoView():document.querySelector(`[id='${decodeURIComponent(e)}']`).scrollIntoView({behavior:"smooth"}),e==="top"?history.replaceState(null,null," "):history.pushState(null,null,`#${e}`)})})</script><script>var mybutton=document.getElementById("top-link");window.onscroll=function(){document.body.scrollTop>800||document.documentElement.scrollTop>800?(mybutton.style.visibility="visible",mybutton.style.opacity="1"):(mybutton.style.visibility="hidden",mybutton.style.opacity="0")}</script><script>document.getElementById("theme-toggle").addEventListener("click",()=>{document.body.className.includes("dark")?(document.body.classList.remove("dark"),localStorage.setItem("pref-theme","light")):(document.body.classList.add("dark"),localStorage.setItem("pref-theme","dark"))})</script><script>document.querySelectorAll("pre > code").forEach(t=>{const n=t.parentNode.parentNode,e=document.createElement("button");e.classList.add("copy-code"),e.innerText="copy";function s(){e.innerText="copied!",setTimeout(()=>{e.innerText="copy"},2e3)}e.addEventListener("click",o=>{if("clipboard"in navigator){navigator.clipboard.writeText(t.textContent),s();return}const e=document.createRange();e.selectNodeContents(t);const n=window.getSelection();n.removeAllRanges(),n.addRange(e);try{document.execCommand("copy"),s()}catch(e){}n.removeRange(e)}),n.classList.contains("highlight")?n.appendChild(e):n.parentNode.firstChild==n||(t.parentNode.parentNode.parentNode.parentNode.parentNode.nodeName=="TABLE"?t.parentNode.parentNode.parentNode.parentNode.parentNode.appendChild(e):t.parentNode.appendChild(e))})</script></body></html>